docker/Dockerfile

```diff
@@ -12,43 +12,267 @@
+RUN pip install --force-reinstall --no-cache-dir PennyLane_Lightning*.whl
+
+# Download Kokkos release and build OpenMP backend
+FROM base-build as build-kokkos-openmp
+WORKDIR /opt/kokkos-${KOKKOS_VERSION}
+RUN wget --progress=dot:giga https://github.com/kokkos/kokkos/archive/refs/tags/${KOKKOS_VERSION}.tar.gz
+RUN tar -zxf ${KOKKOS_VERSION}.tar.gz && mv /opt/kokkos-${KOKKOS_VERSION}/kokkos-${KOKKOS_VERSION}/* /opt/kokkos-${KOKKOS_VERSION}
+RUN cmake -S /opt/kokkos-${KOKKOS_VERSION} -B /opt/kokkos-${KOKKOS_VERSION}/build \
+    -G Ninja \
+    -DCMAKE_BUILD_TYPE=Release \
+    -DCMAKE_INSTALL_PREFIX=/opt/kokkos \
+    -DBUILD_SHARED_LIBS:BOOL=ON \
+    -DBUILD_TESTING:BOOL=OFF \
+    -DKokkos_ENABLE_SERIAL:BOOL=ON \
+    -DKokkos_ENABLE_OPENMP:BOOL=ON \
+    -DCMAKE_CXX_STANDARD=20 \
+    -DKokkos_ENABLE_EXAMPLES:BOOL=OFF \
+    -DKokkos_ENABLE_TESTS:BOOL=OFF \
+    -DKokkos_ENABLE_LIBDL:BOOL=OFF
+RUN cmake --build /opt/kokkos-${KOKKOS_VERSION}/build
+RUN cmake --install /opt/kokkos-${KOKKOS_VERSION}/build
+
+# Download Lightning release and build lightning-kokkos backend with Kokkos-OpenMP
+FROM base-build-python AS build-wheel-lightning-kokkos-openmp
+WORKDIR /opt/pennylane-lightning
+COPY --from=build-kokkos-openmp /opt/kokkos /opt/kokkos
+ENV CMAKE_PREFIX_PATH=/opt/kokkos:$CMAKE_PREFIX_PATH
+RUN wget --progress=dot:giga https://github.com/PennyLaneAI/pennylane-lightning/archive/refs/tags/v${LIGHTNING_VERSION}.tar.gz
+RUN tar -zxf v${LIGHTNING_VERSION}.tar.gz && mv pennylane-lightning-${LIGHTNING_VERSION}/* /opt/pennylane-lightning
+RUN pip uninstall -y pennylane-lightning
+RUN PL_BACKEND=lightning_kokkos python setup.py build_ext
+RUN PL_BACKEND=lightning_kokkos python setup.py bdist_wheel
+
+# Install lightning-kokkos OpenMP backend
+FROM base-runtime AS wheel-lightning-kokkos-openmp
+COPY --from=build-kokkos-openmp /opt/kokkos /opt/kokkos
+COPY --from=build-wheel-lightning-kokkos-openmp /opt/pennylane-lightning/dist/ /
+RUN pip install --force-reinstall --no-cache-dir PennyLane_Lightning*.whl
+
+# Install CUDA-12 in build venv image
+FROM base-build-python as base-build-cuda
+WORKDIR /opt/cuda-build
+RUN curl -o cuda-12_2.run https://developer.download.nvidia.com/compute/cuda/12.2.2/local_installers/cuda_12.2.2_535.104.05_linux.run
+RUN chmod a+x cuda-12_2.run
+RUN ./cuda-12_2.run --silent --toolkit --toolkitpath=/usr/local/cuda-12.2
+ENV PATH=/usr/local/cuda/bin:${PATH}
+ENV LD_LIBRARY_PATH=/usr/local/cuda/lib64:${LD_LIBRARY_PATH}
+
+# Download Kokkos release and build CUDA backend
+FROM base-build-cuda as build-kokkos-cuda
+WORKDIR /opt/kokkos-${KOKKOS_VERSION}
+# RUN wget --progress=dot:giga https://github.com/kokkos/kokkos/archive/refs/tags/${KOKKOS_VERSION}.tar.gz
+# RUN tar -zxf ${KOKKOS_VERSION}.tar.gz && mv /opt/kokkos-${KOKKOS_VERSION}/kokkos-${KOKKOS_VERSION}/* /opt/kokkos-${KOKKOS_VERSION}
+RUN git clone --branch fix-rtld-deepbind https://github.com/cz4rs/kokkos.git /opt/kokkos-${KOKKOS_VERSION}
```

Comment:
What is this version for? Can this be replaced with the official GH repo?

Reply:
> Can this be replaced with the official GH repo?

No, this branch has a bugfix which allows correct execution on CUDA and HIP-supporting devices when building with CUDA-12. I filed an [issue](https://github.com/kokkos/kokkos/issues/6366) about it.

Reply:
I think Kokkos' develop branch now has a fix. I'll double check it works on with both CUDA and HIP and we can switch back to the official Kokkos repo.

Reply:
So the develop commit that works in the CI tests breaks the CUDA/HIP containers. I thus suggest we keep it as `--branch fix-rtld-deepbind https://github.com/cz4rs/kokkos.git` for now.